llvm/test/Transforms/InstCombine/icmp-inttoptr.ll

```diff
@@ -71,8 +63,7 @@ define i1 @inttoptr_size_mismatch(i200 %x, i9 %y) {
 define <2 x i1> @inttoptr_vector_constant_size_mismatch(<2 x i200> %x) {
 ; CHECK-LABEL: @inttoptr_vector_constant_size_mismatch(
 ; CHECK-NEXT:    [[TMP1:%.*]] = trunc <2 x i200> [[X:%.*]] to <2 x i64>
-; CHECK-NEXT:    [[XPTR:%.*]] = inttoptr <2 x i64> [[TMP1]] to <2 x ptr>
-; CHECK-NEXT:    [[CMP:%.*]] = icmp eq <2 x ptr> [[XPTR]], <ptr inttoptr (i9 42 to ptr), ptr inttoptr (i9 123 to ptr)>
+; CHECK-NEXT:    [[CMP:%.*]] = icmp eq <2 x i64> [[TMP1]], <i64 42, i64 123>
```

Comment:
I had to change to `max-iterations=2` because this test case goes through an intermediate stage:
```
<ptr inttoptr (i9 42 to ptr), ptr inttoptr (i9 123 to ptr)>
-->
<i64 ptrtoint (ptr inttoptr (i9 42 to ptr) to i64), i64 ptrtoint (ptr inttoptr (i9 123 to ptr) to i64)>
-->
<i64 42, i64 123>
```
Is that ok?

Reply:
Replacing ConstantExpr::getPtrToInt with use of IRBuilder should fix this.

Reply:
I'm new in this area and could use some more explicit guidance. I _think_ you are suggesting to use something like `return replaceInstUsesWith(ICmp, Builder.CreateICmp(ICmp.getPredicate(), Op0Src, ...))` but I am not clear about what goes in the `...`. Do I need to crack open the `ConstantAggregate` and do something with that?

Reply:
No, I mean to replace `NewOp1 = ConstantExpr::getPtrToInt(RHSC, SrcTy);` with `NewOp1 = Builder.CreateIntToPtr(RHSC, SrcTy);`

The difference is that the IRBuilder will perform DataLayout-aware constant folding.

You could also directly invoke the ConstantFoldCastOperand() API, but going through IRBuilder is preferred in this context.

Reply:
> But even with this, the test case still goes through the intermediate state of `<i64 ptrtoint (ptr inttoptr ...` so it doesn't reach fixpoint in one iteration.

@nikic ping; how do you recommend to proceed? Thanks.

Reply:
I looked into it, and the problem is that ConstantFoldCastInst expects the operand of the ptrtoint to be an inttoptr expression, but here it is a vector of inttoptr. The fold only works after the ptrtoint has been applied element-wise to the vector elements. We should probably fix the constant folding API to make this work directly, but for now you can work around it like this: `ConstantFoldConstant(ConstantExpr::getPtrToInt(RHSC, SrcTy), DL)`

Reply:
Thanks for the pointer!